core/src/test/java/com/linecorp/armeria/client/retry/RedirectRuleTest.java

```diff
@@ -0,0 +1,161 @@
+            });
+
+            sb.service("/https", (ctx, req) -> {
+                final ResponseHeaders headers =
+                        ResponseHeaders.of(HttpStatus.MOVED_PERMANENTLY,
+                                           HttpHeaderNames.LOCATION,
+                                           "https://127.0.0.1:" + server.httpsPort() + "/httpsRedirect");
+                return HttpResponse.of(headers);
+            });
+            sb.service("/httpsRedirect", (ctx, req) -> {
+                assertThat(ctx.sessionProtocol()).isSameAs(SessionProtocol.H2);
+                System.err.println(ctx.sessionProtocol());
+                return HttpResponse.of("httpsRedirection");
+            });
+
+            sb.service("/seeOther", (ctx, req) -> HttpResponse.from(
+                    req.aggregate().thenApply(aggregatedReq -> {
+                        assertThat(aggregatedReq.contentUtf8()).isEqualTo("hello!");
+                        final ResponseHeaders headers = ResponseHeaders.of(HttpStatus.SEE_OTHER,
+                                                                           HttpHeaderNames.LOCATION,
+                                                                           "/seeOtherRedirect");
+                        return HttpResponse.of(headers);
+                    })));
+            sb.service("/seeOtherRedirect", (ctx, req) -> {
+                assertThat(ctx.method()).isSameAs(HttpMethod.GET);
+                return HttpResponse.of("seeOtherRedirection");
+            });
+
+            sb.service("/loop", (ctx, req) -> {
+                final ResponseHeaders headers = ResponseHeaders.of(HttpStatus.MULTIPLE_CHOICES,
+                                                                   HttpHeaderNames.LOCATION, "/loop1");
+                return HttpResponse.of(headers);
+            });
+            sb.service("/loop1", (ctx, req) -> {
+                final ResponseHeaders headers = ResponseHeaders.of(HttpStatus.FOUND,
+                                                                   HttpHeaderNames.LOCATION, "/loop2");
+                return HttpResponse.of(headers);
+            });
+            sb.service("/loop2", (ctx, req) -> {
+                final ResponseHeaders headers = ResponseHeaders.of(HttpStatus.TEMPORARY_REDIRECT,
+                                                                   HttpHeaderNames.LOCATION, "/loop");
+                return HttpResponse.of(headers);
+            });
+        }
+    };
+
+    @BeforeEach
+    void setUp() {
+        counter.set(0);
+    }
+
+    @Test
+    void redirect_successWithRetry() {
+        final AggregatedHttpResponse res = sendRequest(2);
+        assertThat(res.contentUtf8()).contains("fooRedirect");
+    }
+
+    private static AggregatedHttpResponse sendRequest(int totalAttempts) {
+        final RetryConfig<HttpResponse> config =
+                RetryConfig.builder(RetryRule.redirect().orElse(RetryRule.onServerErrorStatus()))
```

Comment:
I still think adding redirect as a retry rule in the API is too complex (implementation can map a ClientOption to the decorator somehow I figure). Have to reason about this `orElse`, should redirect come first, last, does it matter? Seems like a lot of cognitive load for an option that I don't think people generally tie to retry (we tie it since the implementation is similar, but the concept is so different).

Doesn't this seem to lose too much compared to OkHttp, JDK 11, Vert.x, Jetty, all of which followRedirects is a boolean or maybe a simple enum for JDK 11?

Reply:
> implementation can map a ClientOption to the decorator somehow I figure

I was planning to provide something like
`RedirectingClient.newDecorator()`. (I didn't update it yet. Sorry about it. 😅)
Let me see which one is better. 

> Have to reason about this orElse, should redirect come first, last, does it matter?

It doesn't matter. It's just showing that we can use the redirection with retry together.
Because of that, I think I still need `RetryRule.redirect()`. If we don't have, we have to add `RetryingClient` twice which has some overhead.

Reply:
I still think providing it as a decorator is a bit better because a user can easily figure out when he/she put the decorators in the wrong order:
```java
// Logs all the requests and responses
WebClient.builder(...)
         .decorator(LoggingClient.newDecorator())
         .decorator(RedirectingClient.newDecorator())
         .build()

// Logs the first request and last response.
WebClient.builder(...)
         .decorator(RedirectingClient.newDecorator())
         .decorator(LoggingClient.newDecorator())
         .build()
```
If we provide it as an option, it's a bit hard to manipulate the result. 🤔

Reply:
Had a chat with @trustin and @ikhoon and I decided not to use a decorator for redirecting. 
Please stay tuned. 😄